File: py_v/battles.py
from utils import *
from characters import *


class Battle(object):
    logger = GameLogger(GameLogger.LOG_FILE["battle"])
    num_battles = 0

    @staticmethod
    def tick(logging=False):
        if logging and Battle.num_battles is 0:
            Battle.logger.clean()
        Battle.num_battles += 1

    def __init__(self, party_1, party_2, logging=False):
        Battle.tick(logging)

        if not isinstance(party_1, list):
            party_1 = [party_1]
        if not isinstance(party_2, list):
            party_2 = [party_2]

        self.party_1 = party_1
        self.party_2 = party_2

        self.is_logging = logging
        if self.is_logging:
            Battle.logger.write("\nBattle.__init__(): Start Logging Battle #%3d" % (
                Battle.num_battles))

    def is_active(self, battle_queue):
        party_1_active = 0
        party_2_active = 0

        for member in self.party_1:
            if member in battle_queue and member.is_alive():
                party_1_active += 1

        if party_1_active < 1:
            if self.is_logging:
                Battle.logger.write(
                    "  Battle.is_active(): no one in party_1 - battle over")
                return False

        for member in self.party_2:
            if member in battle_queue and member.is_alive():
                party_2_active += 1

        if party_2_active < 1:
            if self.is_logging:
                Battle.logger.write(
                    "  Battle.is_active(): no one in party_2 - battle over")
                return False
        return True

    def run(self):
        battle_queue = self.party_1 + self.party_2
        battle_queue.sort(key=lambda member: member.speed)

        while(self.is_active(battle_queue)):
            for member in battle_queue:
                if member.is_alive():
                    member.engage(battle_queue)

File: rb_v/battles.rb
require './utils.rb'
require './characters.rb'

class Battle
  LOGGER = GameLogger.new(GameLogger::LOG_FILE[:battle])
  @@num_battles = 0

  def self.tick(logging = false)
    LOGGER.clean if logging && @@num_battles == 0
    @@num_battles += 1
  end

  def initialize(party_1, party_2, logging = false)
    Battle.tick(logging)

    party_1 = [party_1] unless party_1.instance_of?(Array)
    party_2 = [party_2] unless party_2.instance_of?(Array)

    @party_1 = party_1
    @party_2 = party_2

    @is_logging = logging
    if @is_logging
      LOGGER.write("\nBattle.initialize: Start Logging Battle #{@@num_battles}")
    end
  end

  def is_active?(battle_queue)
    party_1_active = 0
    party_2_active = 0

    @party_1.each do |member|
      party_1_active += 1 if battle_queue.include?(member) && member.is_alive?
    end
    if party_1_active < 1
      LOGGER.write('  Battle.is_active?: no one in party_1 - battle over') if @is_logging
      return false
    end

    @party_2.each do |member|
      party_2_active += 1 if battle_queue.include?(member) && member.is_alive?
    end
    if party_2_active < 1
      LOGGER.write('  Battle.is_active?: no one in party_2 - battle over') if @is_logging
      return false
    end

    true
  end

  def run
    battle_queue = @party_1
    battle_queue += @party_2
    battle_queue = battle_queue.sort_by(&:speed)

    while is_active?(battle_queue)
      battle_queue.each do |member|
        member.engage(battle_queue) if member.is_alive?
      end
    end
  end
end
